RRS-148 create server action to add property ownership

OttoChu opened the pull request:
I am waiting for the landlord profile stuff to be done before changing the dates

OttoChu committed:
Added basic claim property actions

LukeHosk committed:
tutorial notes

OttoChu committed:
Started to add checks

OttoChu committed:
Completed the checks to claim property

I have just throw errors when the property has already been claimed. Might need to change it later once the landlord table stuff is done

LukeHosk committed:
Merged origin/development into RRS-148-create-server-action-to-add-property-ownership

LukeHosk commented:
I'm going to add a quick form to the page you made so we can test the function works correctly

LukeHosk committed:
Merged origin/development into RRS-148-create-server-action-to-add-property-ownership

LukeHosk committed:
create basic property claiming form for testing

LukeHosk committed:
update function definition to:
- use bound values for the IDs
- correctly accept and return the state

LukeHosk committed:
use validation on action inputs & use Date instead of string for comparisons

LukeHosk committed:
extend validation to ensure dates are in the past

LukeHosk committed:
Issue identified: function does not return a state on all paths.

start to fix by adding a success message after adding data

LukeHosk committed:
extract data setting to seperate function

LukeHosk committed:
removed if-not-null check on data that is already guarunteed not to be null.

set property ownership immediatly if no existing ownership records are found

LukeHosk committed:
planning: define plan for claim checking

LukeHosk committed:
move ownership record creation from within loop to after (function would have tried to make a new record for every existing claim it didn't overlap).

Create function to close an open claim before creating the new one

allow action to track if there is an open claim and call the function created above to close it if there is

LukeHosk committed:
set claim creation to use `insert` not `update`

LukeHosk committed:
landlord_id should not be nullable

LukeHosk committed:
style: remove un-necessary else-ifs after conditions that cause the function to return. flattens code and increases readability.

use standard formatting for all return statements

LukeHosk commented:
when trying to test this, it did not work at all.

it should now be in a state where it can work and be tested even though I don't think the logic is complete, but you should be able to fix that without me

LukeHosk commented on app/properties/claim/actions.ts:
```diff
@@ -0,0 +1,237 @@
+        };
+
+    //  check a the user is already the landlord of the property for this time period
+    const { data: propertyOwnershipList, error: propertyOwnershipError } = await supabase
+        .from('property_ownership')
+        .select('landlord_id, started_at, ended_at')
+        .eq('property_id', property_id)
```
there are currently no rls policies on this table to allow selection.

we either want to add them (you can remove or change the migration I added since we have not yet pushed it to development/production)

or you can use a service supabase client like I do in lines 181-193 to overwrite RLS

I'd recommend the first option since we want this table to be readable anyway, but we can do either for now

LukeHosk commented on app/properties/claim/actions.ts:
```diff
@@ -0,0 +1,237 @@
+            ended_at
+        );
+
+    // If not empty, for each existing property ownership record, there several cases to consider:
+    // 1. The existing claim is closed
+    //    1.1 The existing claim does not overlap with the new claim, continue
+    //    1.2 The existing claim overlaps with the new claim, fail
+    // 2. The existing claim is open 
+    //    2.1 The new claim starts after the existing claim, close the existing claim & continue
+    //    2.2 The new claim starts before the existing claim
+    //        2.2.1 The new claim is closed and ends before the existing claim starts, continue
+    //        2.2.2 The new claim is open or ends after the existing claim starts, fail
+
+    // To handle 2.1, track whether there is an open claim to close as we iterate through the existing claims
+    let openClaimToClose: false | { started_at: Date } = false;
+
+    for (const propertyOwnership of propertyOwnershipList) {
+        const existing_start = new Date(propertyOwnership.started_at)
+        const existing_end = propertyOwnership.ended_at ? new Date(propertyOwnership.ended_at) : new Date()
+
+        // tried to claim the same property again
+        if (propertyOwnership.landlord_id == user.id)
+            return {
+                message: 'User is already the landlord of this property'
+            };
+
+        // new start date is before current end date
+        if (existing_end && ended_at < existing_end)
+            return {
+                message: 'New start date is before the current end date'
+            };
+
+        // new end date is after current start date
+        if (existing_start && started_at > existing_end)
+            return {
+                message: 'New end date is after the current start date'
+            };
+
+    }
```
I believe what we are trying to do here is what I outlined in the comment here

I changed the layout of this code block from what you had but I don't believe I changed much if any of the logic outside of moving the adding of ownership to after the loop.

after making it more readable it seems like it does not do what we want (it also doesn't do what you say it does in the comments, your 'new start date is before current end date' check only references end dates).

Can you check whether my test plan in the comments is what we want, and if it is, adjust this section to match it

LukeHosk reviewed (changes requested):
as detailed in comments above:

1. ensure function has permission to fetch property_ownership records
2. check logic for overlap checks

OttoChu committed:
fixed it

OttoChu committed:
actually fixed it

OttoChu committed:
Merge branch 'development' into RRS-148-create-server-action-to-add-property-ownership

OttoChu committed:
Update supabase.types.ts

LukeHosk committed:
fix index issue
